### PR TITLE
MINOR: after reading BYTES type it's possible to access data beyond its size

### PR DESCRIPTION
**Pull request by bachmanity1**
After reading data of type `BYTES`, `COMPACT_BYTES`, `NULLABLE_BYTES` or `COMPACT_NULLABLE_BYTES` returned `ByteBuffer` might have a capacity that is larger than its limit, thus these data types may access data that lies beyond its size by increasing limit of the returned `ByteBuffer`. I guess this is not very critical but I think it would be good to restrict increasing limit of the returned `ByteBuffer` by making its capacity strictly equal to its limit.  I think someone might unintentionally mishandle these data types and accidentally mess up data in the `ByteBuffer` from which they were read. 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by bachmanity1** [comment]
@showuon @C0urante could you please review this? Thanks.

**Comment by C0urante** [comment]
Hi @bachmanity1! This is a bit outside my area of expertise and it's probably best for another committer to review.

**Comment by bachmanity1** [comment]
@ijuma could you please review this? Thanks!

**Comment by bachmanity1** [comment]
@cmccabe could you please review this? Thanks!

**Review by showuon — APPROVED** [review]
LGTM! Thanks for the patch!

File: clients/src/main/java/org/apache/kafka/common/protocol/types/Type.java
Comment (showuon) [inline clients/src/main/java/org/apache/kafka/common/protocol/types/Type.java]: nit: Since we will use `buffer.position() + size` twice, should we make it as a variable?

**Review by showuon — COMMENTED** [review]
One minor comment left. Thanks.

File: clients/src/main/java/org/apache/kafka/common/protocol/types/Type.java
Comment (bachmanity1) [reply clients/src/main/java/org/apache/kafka/common/protocol/types/Type.java]: I introduced a new variable `newPosition`

**Review by showuon — APPROVED** [review]
LGTM! Let's wait for the CI build completed

**Comment by bachmanity1** [comment]
The CI test failures look unrelated to this PR

**Comment by showuon** [comment]
Failed tests are unrelated
```
    Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsWithCustomForwardingAdminIntegrationTest.testReplicationIsCreatingTopicsUsingProvidedForwardingAdmin()
    Build / JDK 8 and Scala 2.12 / kafka.admin.ReassignPartitionsIntegrationTest.testReassignment(String).quorum=kraft
    Build / JDK 17 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.DedicatedMirrorIntegrationTest.testMultiNodeCluster()
    Build / JDK 17 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsWithCustomForwardingAdminIntegrationTest.testCreatePartitionsUseProvidedForwardingAdmin()
    Build / JDK 17 and Scala 2.13 / kafka.server.KafkaServerKRaftRegistrationTest.[1] Type=ZK, Name=testRegisterZkBrokerInKraft, MetadataVersion=3.4-IV0, Security=PLAINTEXT
```

**Pull request merged**
